### ISSUE TITLE
Outline sidebar

### ISSUE THREAD

**Issue by tobimensch** [fox0430/moe#935]
Many editors offer a feature that is often called minimap, but I think while that looks cool, and it's basically a fancy version of a scrollbar, it doesn't really help to understand the structure of a file/document all that much.

However, the table of contents sidebar, that is offered in many PDF viewers for PDF files that have a table of contents, makes a lot of sense and actually helps to navigate a long document.

I would like a sidebar, preferably on the left by default, that contains an outline(table of contents) of the current buffer's content.

This would, of course, depend on the type of file/document. I'm not suggesting here to make moe into a PDF viewer (for now). But for example for Markdown and Asciidoc documents, and also many different kinds of source files, it's possible to generate such an outline. Either there're headings and subheadings, or there are type definitions and methods/functions (sometimes nested similar to headings).

This would make it easier to jump between different parts of a document/source file.

Here's a extremely simplified mockup of what I'm proposing:
```
import section         |100  proc openWindow(w: Window) =
type Window            |101    proc butWhy() =
type Door              |102      echo "but Why?"
type Chair             |103    butWhy()
type Wall              |104    w.opened = true
proc openWindow        |105
L proc butWhy          |106  proc openDoor(d: Door) =
proc openDoor          |107    d.opened = true
proc sitOnChair        |108
proc stareAtWall       |109  proc sitOnChair(c: Chair) =
proc closeWindow       |110    c.isUsed = true
proc closeDoor         |111
```

This proposed outline can be used to navigate the document using the cursor
and the `Enter` or `Space` key.

I'd also like there to be a *preview* feature in that outline, that shows you
the content of a part of the document, without changing the scroll position
of the underlying buffer. This would be useful to take a glance at something,
without having to interrupt the workflow too much.

Another feature that outline should have is to open a the part of the file/document the cursor is on in a new horizontal or vertical split. 

But the most important aspect of this idea is that it should work with the same
key combinations, and the same general navigational concept, across different
types of source files and documents. So it doesn't matter if you're viewing
a C++ file, a Markdown document, a Man page or whatever else, you can always
rely on this outline feature working in the exact same way.

Last but not least, typical vim navigational features should also work in this sidebar. For example searching with `/`.

To sum it up:
- Sidebar (configurable either on the left or on the right)
- Can be used to navigate to different parts of any type of document/source file
- Preview feature for seeing a part of a document, without actually navigating there
- Understands nested structures
- Vim like navigation
